src/utp_stream.cpp

```diff
@@ -2055,21 +2055,35 @@ bool utp_socket_impl::resend_packet(packet* p, bool fast_resend)
 		TORRENT_ASSERT(!m_stalled);
 		m_stalled = true;
 		m_sm.subscribe_writable(this);
-		p->need_resend = true;
+		if (!p->need_resend)
+		{
+			p->need_resend = true;
+			m_needs_resend.push_back(p);
+		}
 		m_bytes_in_flight -= p->size - p->header_size;
 	}
-	else if (ec)
-	{
-		m_error = ec;
-		set_state(state_t::error_wait);
-		test_socket_state();
-		return false;
-	}
 	else
 	{
-		m_sm.inc_stats_counter(counters::utp_packets_out);
-		++m_out_packets;
-		++p->num_transmissions;
+		if (p->need_resend)
+		{
+			p->need_resend = false;
+			auto entry = std::find(m_needs_resend.begin(), m_needs_resend.end(), p);
```

Comment:
this looks like it could be a problem under edge-case conditions. But probably not worse that the current implementation

Reply:
Yeah, it could get bad if there are a lot of packets in flight, although I've only observed at most ~70 packets in the buffer at once due to the uTP max window size.

This could probably also be improved by using an intrusive linked list.